dali/driver/hasseb.py

```diff
@@ -1,128 +1,193 @@
                 # 2: "OK"
+                return BackwardFrame(data[1])
+            elif response_status == HASSEB_DRIVER_INVALID_ANSWER:
                 # 3: "Invalid Answer"
-                responseStatus = rdData[0]
-                
-                if responseStatus == 2:
-                    return dali.frame.BackwardFrame(rdData[1])
-                elif responseStatus == 3:
-                    return dali.frame.BackwardFrameError(255)
-                
-            return None
-
-
-    def send(self, command):
-        if self.ep is None:
-            self._openDevice()
-
-        assert isinstance(command, Command)
-        
-        needsResponse = command.response is not None
-        a, b = command.frame.as_byte_sequence
-        
-        # print(u" SEND: a=0x%x b=0x%x"%(a,b))
-        
-        backward = self._writeDali( a, b, needsResponse)
-        if command._response:
-            response = command._response(backward)
+                return BackwardFrameError(255)
+        self.logger.error("Invalid Frame")
+        return None
+
+class SyncHassebDALIUSBDriver(HassebDALIUSBDriver, SyncDALIDriver):
+    """Synchronous ``DALIDriver`` implementation for Hasseb DALI USB device.
+    """
+
+    def __init__(self, bus=None, port_numbers=None, interface=0):
+        self.backend = USBBackend(
+            HASSEB_USB_VENDOR,
+            HASSEB_USB_PRODUCT,
+            bus=bus,
+            port_numbers=port_numbers,
+            interface=interface
+        )
+
+    def send(self, command, timeout=2000):
+        self.backend.write(self.construct(command))
+        frame = None
+        backoff = 0.010
+
+        if command.response is not None:
+            for i in range(7):
+                frame = self.extract(self.backend.read(timeout=timeout))
+                if isinstance(frame, HassebDALIUSBNoAnswer):
+                    self.backend.write(self.construct(command))
+                if isinstance(frame, BackwardFrame):
+                    if command.response:
+                        return command.response(frame)
+                    return frame
+                backoff += backoff
+                sleep(backoff)
+            raise MissingResponse()
+        return None
+
+class AsyncHassebDALIUSBDriver(HassebDALIUSBDriver, AsyncDALIDriver):
+    """Asynchronous ``DALIDriver`` implementation for Hasseb DALI USB device.
+       This driver is FAKE, since Hasseb does NOT support async communication.
+    """
+    _pending = None
+
+    def __init__(self, bus=None, port_numbers=None, interface=0):
+        self.backend = USBListener(
+            self,
+            HASSEB_USB_VENDOR,
+            HASSEB_USB_PRODUCT,
+            bus=bus,
+            port_numbers=port_numbers,
+            interface=interface
+        )
+
+    def send(self, command, callback=None, **kw):
+        data = self.construct(command)
+        if command.response is not None:
+            self._pending = command, callback, kw
```

Comment:
A more robust implementation would imo be a fifo queue for pending requests. This way, in theory some can send multiple requests at once. Not sure if hasseb firmware can handle this, needs to be checked. If firmware requires one request/response cycle to be finished before processing next one, at least an exception would be appropriate pointing out that theres an already pending request.

Reply:
Firmware can't handle multiple requests. 

Implementing a queue would require a background thread to consume the queue elements or a true async implementation (using asyncio).

Throwing an exception is not an option, sometime commands fail with this controller and you would be locked (commands fail quite often with this controller, that why I added the retry algorithm with backoff)

Reply:
the firmware waits for the answer itself... if the answer hasn't arrived yet you receive the NoAnswer frame that I'm parsing already